### PR TITLE
[SPARK-50241][SQL] Replace NullIntolerant Mixin with Expression.nullIntolerant method

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?
Replace NullIntolerant Mixin with Expression.nullIntolerant method

### Why are the changes needed?

https://github.com/apache/spark/pull/48758#issuecomment-2458713378 via @cloud-fan 
> This is not the first time that we are restricted by the trait-based tagging system. Extending a trait is static as it happens at compile but we really want to be more dynamic at runtime. For example, we added Expression#stateful and removed the trait StatefulExpression a while ago. I think the same thing happens to NullIntelerant now. Shall we also add Expression#nullIntolerant?

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

Modified ExpressionInfoSuite,
- Check whether SQL expressions should extend NullIntolerant

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @yaooqinn .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 1]
```diff
@@ -322,7 +322,8 @@ case class StaticInvoke(
     propagateNull: Boolean = true,
     returnNullable: Boolean = true,
     isDeterministic: Boolean = true,
-    scalarFunction: Option[ScalarFunction[_]] = None) extends InvokeLike {
+    scalarFunction: Option[ScalarFunction[_]] = None,
+    override val nullIntolerant: Boolean = false) extends InvokeLike {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:326]: This flag is the same as `propagateNull`?

BTW I kind of like the name `propagateNull` more as it's more straightforward. But I'm not a native English speaker and I could be wrong. cc @dtenedor @srielau

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:326]: I guess they are the same too. If you don't mind， I'd like too keep the PR in AS-IS state to minimize the potential influence of this refactoring work. Since nullIntolerant defaults to false，propagateNull defaults true，mapping them blindly might result in many unsure plan changes，I‘d like to revist and revise them in followups in the meanwhile I'm fixing the RuntimeReplaceables which use StaticInvoke and Invoke heavily？ WDYT？ @cloud-fan

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:326]: nullIntolerant and propagateNull currently have opposite defaults in where they are defined and the places they are called, I guess we shall leave them AS-IS and change them more carefully

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:326]: > OK, but at lease we shouldn't add a new parameter here. We can just add
> 
> ```
> override def nullIntolerant = propagateNull
> ```

I have tested this change in a separate push, the actions did show some test failures to me. I suggest we do that change separately.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:326]: Does this mean `propagateNull` and `nullIntolerant` are not exactly the same? I'd like to figure this out now, instead of adding a new parameter to `StaticInvoke` (not binary compatible) and then remove it later.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:326]: I didn't mean that, I mean propagateNull lacks of some optimizations applied to the NullIntolerant trait

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:326]: So it's not even related to `Invoke`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -549,9 +552,11 @@ trait StringBinaryPredicateExpressionBuilderBase extends ExpressionBuilder {
 
 object BinaryPredicate {
   def unapply(expr: Expression): Option[StaticInvoke] = expr match {
-    case s @ StaticInvoke(
-        clz, _, "contains" | "startsWith" | "endsWith", Seq(_, _), _, _, _, _, _)
-      if clz == classOf[ByteArrayMethods] => Some(s)
+    case s: StaticInvoke
+        if s.staticObject == classOf[ByteArrayMethods] &&
+          Set("contains", "startsWith", "endsWith").contains(s.functionName) &&
+          s.arguments.length == 2 =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:558]: why is this change?

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala [hunk 3]
```diff
@@ -288,12 +288,13 @@ class ExpressionInfoSuite extends SparkFunSuite with SharedSparkSession {
       candidateExprsToCheck.filter(superClass.isAssignableFrom).foreach { clazz =>
         val isEvalOverrode = clazz.getMethod("eval", classOf[InternalRow]) !=
           superClass.getMethod("eval", classOf[InternalRow])
-        val isNullIntolerantMixedIn = classOf[NullIntolerant].isAssignableFrom(clazz)
+        val isNullIntolerantMixedIn = clazz.getMethod("nullIntolerant") !=
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ExpressionInfoSuite.scala:291]: does it mean `overrideNullIntolerant`?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala [hunk 4]
```diff
@@ -1633,15 +1633,15 @@ class JDBCV2Suite extends QueryTest with SharedSparkSession with ExplainSuiteHel
...
-      "aes_encrypt(cast(null as string), name) is null")
+      "aes_encrypt(name, '1234567812345678') != 'spark'")
     checkFiltersRemoved(df5, false)
-    val expectedPlanFragment5 = "PushedFilters: [], "
+    val expectedPlanFragment5 = "PushedFilters: [NAME IS NOT NULL], "
     checkPushedInfo(df5, expectedPlanFragment5)
     checkAnswer(df5, Seq(Row("amy"), Row("cathy"), Row("alex"), Row("david"), Row("jen")))
 
     val df6 = sql("SELECT name FROM h2.test.employee WHERE " +
       "aes_decrypt(cast(null as binary), name) is null")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala:1643]: can we change the query to not use `null` input? We can use `X'...'` to write binary literal.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 5]
```diff
@@ -805,11 +807,13 @@ object RegExpExtractBase {
...
 
 abstract class RegExpExtractBase
-  extends TernaryExpression with ImplicitCastInputTypes with NullIntolerant {
+  extends TernaryExpression with ImplicitCastInputTypes {
+  override def nullIntolerant: Boolean = true
   def subject: Expression
   def regexp: Expression
   def idx: Expression
 
+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:816]: unnecessary change

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala (hunk 4 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala:1643]: Well, the second point makes me consider a potential breaking change: null intolerance takes higher precedence for error-raising in certain cases.

**Comment by yaooqinn** [comment]
Merged to master. Thank you for the review @dongjoon-hyun @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala:1643]: Does this mean `aes_decrypt` should not set `propagateNull` to true?

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala:1643]: I haven't done a full check, but I believe some parameter checks may occur in the original expressions, while others might be checked in the replacement, which could result in different behaviors.
